Redesign jQuery UI datepicker

pixelipo opened the pull request:
Fixes #5691 and nextcloud/deck#205
Before:
![image](https://user-images.githubusercontent.com/4129927/28168264-d104535c-67de-11e7-8be1-274be86d8117.png)

After:
![image](https://user-images.githubusercontent.com/4129927/28168320-fd007f80-67de-11e7-8897-09131f35ec06.png)

Tested in Files, Calendar and Deck apps.

Should also be considered regarding nextcloud/calendar#533

I've deliberately targeted datepicker only, since there are already some jQuery UI fixes in place in several files.

pixelipo committed:
Redesign jQuery UI datepicker

Signed-off-by: Marin Treselj <marin@pixelipo.com>

georgehrke commented:
I think the weekends should be greyed out (just like in Jan's mockup) and not colored in the theme color. Using the theme-color highlights them and makes them seem more important, which they are not in most cases.

Why is 18 highlighted? Is that the one you are currently hovering? 

In the calendar we use a light yellow to highlight the current day in the main calendar grid. What about using it in the datepicker too? :)

pixelipo commented:
@georgehrke I have tried greying out weekends, but it was either not discernible, or it lacks contrast (accessibility concerns).

Yes, 18th is `:hover`. Yellow doesn't exist anywhere in the NC design - I wouldn't add colts which are not in the design "spec". Theme primary is used as highlight colour in plenty of places across Nextcloud.

georgehrke commented:
Hm. Do you have these concerns for the colors chosen by Jan too? 
![datepicker](https://user-images.githubusercontent.com/925062/28124080-f4406f98-6722-11e7-92ab-529e542a3582.png)

Espina2 commented:
>  I think the weekends should be greyed out (just like in Jan's mockup) and not colored in the theme color. Using the theme-color highlights them and makes them seem more important, which they are not in most cases.

I agree with @georgehrke 

Removing the first row with days of the week is step a back, the sligh whiter grey can signify a lot of things. I also think it will look more clean if a more generous margins in all the sizes.
@jancborchardt

pixelipo commented:
To tell you the truth, I didn't even notice that weekends were lighter than the test on Jan's mockup until you mentioned it. I had to reopen the image and squint to notice it. Regular user with a not-Eizo monitor will never see it. So, I don't agree with that part. 
I do agree with the current-day highlight colour, which I implemented.

Re paddings, I think that might be a good idea. 

I haven't removed the row with names of the weekdays.

jancborchardt commented:
Nice! Just some details:

- Highlight for selected date should be bild text.
- Hover highlight for day should be the same as for the selected date. We shouldn't have too many different states.
- Can you remove hover effect on the week? It's too mich noise, and also doesn't work on mobile anyway.
- Row for days of the week should indeed stay. But non-bold and at 50% opacity so it doesn't take away the focus.
- Can you make the day highlights round as in my mockup?
- Dates of previous and next month could be a bit lighter?
- I agree with the concerns about marking the weekend blue, making it too present instead of subdued. Try using the same grey-tone / opacity-level as for the months previous/after maybe?
- The box should be centered below the date field, with triangle in the middle.

jancborchardt commented:
Oh another thing, also related to the highlights being circles: can you make sure each day / clickable element is 44*44px so it's properly tappable on mobile? :)

pixelipo commented:
@jancborchardt 

>    Highlight for selected date should be bold text.

:+1:

>    Hover highlight for day should be the same as for the selected date. We shouldn't have too many different states.

:+1:

>    Can you remove hover effect on the week? It's too much noise, and also doesn't work on mobile anyway.

:+1:

>    Row for days of the week should indeed stay. But non-bold and at 50% opacity so it doesn't take away the focus.

:+1:

>    Can you make the day highlights round as in my mockup?

As soon as you show me a single example where round border/background is used in Nextcloud :wink:  I used 3px border-radius which seems to be the default

>    Dates of previous and next month could be a bit lighter?

I can try, but we are getting close to the accessibility issues here - think bad monitors and people with bad eyesight. There is a nice [webapp](#D2D3DA) for calculating contrast and your design breaks Accessibility Guidlines

>    I agree with the concerns about marking the weekend blue, making it too present instead of subdued. Try using the same grey-tone / opacity-level as for the months previous/after maybe?

Seems like there is a general consensus about this so I'll change this even though I disagree. Most printed calendars highlight weekends and it doesn't affect finding the correct day, as long as that highlight is different from the selected/today highlights. I wonder what the UX tests would 

>    The box should be centered below the date field, with triangle in the middle.

:+1: 

>    can you make sure each day / clickable element is 44*44px so it's properly tappable on mobile? :)

I gave this a quick shot and it makes the whole popover HUGE - 314 px wide (more if I add suggested padding around edges...this would affect usability on smaller screens... So I'm not sure it's feasible...

pixelipo committed:
Fixes issues mentioned in comments.

- [x] Highlight for selected date should be bold text.
- [x] Hover highlight for day should be the same as for the selected date.
- [x] Remove hover effect on the week.
- [x] Row for days of the week should indeed stay, but non-bold and at 50% opacity so it doesn't take away the focus.
- [x] Dates of previous and next months a bit lighter.
- [x] Remove marking the weekend blue.
- [x] The box centered below the date field, with triangle in the middle.

Signed-off-by: Marin Treselj <marin@pixelipo.com>

pixelipo committed:
Add padding to popover

Signed-off-by: Marin Treselj <marin@pixelipo.com>

pixelipo commented:
- [x] Highlight for selected date should be bold text.
- [x] Hover highlight for day should be the same as for the selected date.
- [x] Remove hover effect on the week.
- [x] Row for days of the week should indeed stay, but non-bold and at 50% opacity so it doesn't take away the focus.
- [x] Dates of previous and next months a bit lighter.
- [x] Remove marking the weekend blue.
- [x] The box centered below the date field, with triangle in the middle.
- [x] Add more padding to the calendar

Result (11 is selected, 19 is hovered):
![image](https://user-images.githubusercontent.com/4129927/28213631-3d5aaf92-68a7-11e7-921e-e3ec8ad9412a.png)


I used `color: nc-lighten($color-main-text, 33%)` for weekends, which passes the [contrast accessibility test](https://leaverou.github.io/contrast-ratio/#%23545454-on-white). previous/next month dates have additional `opacity: .7` set in jquery-ui.css and therefore [fails the test](https://leaverou.github.io/contrast-ratio/#%238D8D8D-on-white). Should I remove it?

pixelipo committed:
Remove deprecation

Signed-off-by: Marin Treselj <marin@pixelipo.com>

georgehrke commented:
Does "July 2017" have the same color as the ordinary days?
It looks to prominent imho

pixelipo commented:
@georgehrke it's bold, that's default jquire-ui.css - but I agree that normal would look better.

jancborchardt commented:
Yup, agree not bolding the month would be better. :) But please bold the selected/hovered date.

> As soon as you show me a single example where round border/background is used in Nextcloud 😉

@pixelipo true, it's not used that much, but in this case it looks much better. Two examples are 1) avatars / avatar placeholders and 2) radio buttons, which this is kinda very similar to.

Agree with the assessment about the huge size re 44*44px so let's leave that out for now.

pixelipo committed:
Make dates slightly larger and perfectly rounded.

SCSS cleanup, fix fringe cases, add margin between dates, un-bold title.

Signed-off-by: Marin Treselj <marin@pixelipo.com>

pixelipo commented:
Ok, I think this is ready for final review and merging now. I've fixed several things in the last commit:

- [x] dates are now completely rounded
- [x] cells are perfect squares (so highlights are not ovals, but perfect circles)
- [x] dates are a couple of pixels larger (better usability on mobile, cleaner UI, helps with :arrow_up: )
- [x] added 1px margin around dates (looks better when selected/hover/today are next to each other - see screenshot)
- [x] month name is not bold anymore
- [x] selected date is bold (actually, it always was, but it's not that noticeable)
- [x] fixed some fringe cases (hovering over today when selected, selected is in another month etc.)

jancborchardt commented:
Just one last thing: Weekday names in header should also be same grey tone as the "previous month" color. They don't need to be that present.

Othetwisr looks great, and I'll be on vacation so can't review more. ;)

jancborchardt commented:
Then the lightest grey which is still accessible. :) The weekday names should not stand out as much as the day numbers themselves.

(Btw thanks for your great work and for checking these things! We really need to improve also on keyboard accessibility. :)

pixelipo committed:
Make weekday names brighter, lighter month/year title.

Signed-off-by: Marin Treselj <marin@pixelipo.com>

pixelipo committed:
Fix font-weight for datepicker title.

Signed-off-by: Marin Treselj <marin@pixelipo.com>

pixelipo commented:
- [x] weekday names are now lightest [accessible](https://leaverou.github.io/contrast-ratio/#rgba%2884%2C84%2C84%2C.8%29-on-white) grey. We might want to consider adding this color as a variable in SCSS.
- [x] month/year title is even less bold than before (`font-weight: 300` seems to be NC default
- [x] weekends have the same light grey tone as non-current-month dates

This is as good as it gets, folks. Please review and (finally) merge :wink: 

Files:
![image](https://user-images.githubusercontent.com/4129927/28453157-4dcb8fb4-6df6-11e7-873d-282e83d37b11.png)

Calendar:
![image](https://user-images.githubusercontent.com/4129927/28453121-1f0748f8-6df6-11e7-8e94-16dc0760a73f.png)

Deck:
![image](https://user-images.githubusercontent.com/4129927/28453135-3d31d79e-6df6-11e7-92df-84532fde3eec.png)

pixelipo commented:
cc @nextcloud/tasks might want to deprecate their own customizations of this modal (after this is merged)

cc @nextcloud/accessibility for review

MorrisJobke commented:
> This is as good as it gets, folks. Please review and (finally) merge 😉

😳 Let me have a look at this

raimund-schluessler commented:
> cc @nextcloud/tasks might want to deprecate their own customizations of this modal (after this is merged)

Yes, we want. 😄

rullzer commented on lib/private/legacy/template.php:
```diff
@@ -108,9 +108,9 @@ public static function initTemplateEngine($renderAs) {
 				}
 			}
 
+			OC_Util::addStyle('server', null, true);
```
Why this change?

pixelipo replied:
@rullzer change the order in which styles are loaded. This way, our styles are overriding `jquery-ui` ones by default.

skjnldsv commented:
Amazing work guys!! Really fit the nextcloud overall design!! 😍

rullzer commented:
Rebased upon mater to trigger CI
But otherwise lets get this in!

MariusBluem reviewed (approved):
Tested & works 👍

MariusBluem commented:
🎉🎉🎉 Thanks to everyone who contributed 🙈 Great work 👍

skjnldsv commented:
Well done!!

jancborchardt commented:
I was already on vacation without internet. But have nothing to add @pixelipo – awesome work! :) Hope to have you at the Nextcloud conf end of August in Berlin! https://nextcloud.com/conf/ – there’s travel funding for awesome contributors like you, and you should give a lightning talk about your work. :)

(Same goes for cool folks like @skjnldsv @juliushaertl @Espina2 @raimund-schluessler etc. as usual ;)
